packages/falso/src/lib/core/core.ts

```diff
@@ -11,21 +12,130 @@ type Return<T, O extends FakeOptions> = [O] extends [never]
   : T;
 
 export function fake<T, Options extends FakeOptions>(
-  data: T[] | ((i: number) => T),
-  options?: Options
+  data: T[] | (() => T),
+  options?: Options,
+  comparisonFunction: (item: T, items: T[]) => boolean = primitiveValueIsUnique,
+  comparisonKeys?: string[]
 ): Return<T, Options> {
-  const dataSource = Array.isArray(data) ? () => randElement(data) : data;
+  if (options?.length === 0) {
+    return [] as any;
+  }
 
-  if (options?.length === undefined) {
-    return dataSource(0) as any;
+  if (Array.isArray(data)) {
+    return fakeFromArray(data, options) as any;
   }
 
-  return Array.from({ length: options.length }, (_, index) =>
-    dataSource(index)
+  return fakeFromFunction(
+    data,
+    comparisonFunction,
+    comparisonKeys,
+    options
   ) as any;
 }
 
-export function randElement<T>(arr: T[]) {
+export function fakeFromFunction<T, Options extends FakeOptions>(
+  data: () => T,
+  isItemADuplicateFunction: (
+    item: T,
+    items: T[],
+    comparisonKeys?: string[]
+  ) => boolean,
+  comparisonKeys?: string[],
+  options?: Options
+): T | T[] {
+  if (!options?.length) {
+    return data();
+  }
+
+  const priority = options?.priority ?? 'length';
+
+  if (priority === 'length') {
+    return Array.from({ length: options.length }, (_, index) => data());
+  }
+
+  const items: T[] = [];
+
+  let attempts = 0;
+  const maxAttempts = options.length * 2;
+
+  while (items.length < options.length && attempts < maxAttempts) {
+    const item = data();
+
+    if (!isItemADuplicateFunction(item, items, comparisonKeys)) {
+      items.push(item);
+    }
+
+    attempts++;
+  }
+
+  return items;
+}
+
+export function fakeFromArray<T, Options extends FakeOptions>(
+  data: T[],
+  options?: Options
+): T | T[] {
+  if (!options?.length) {
+    return randElement(data);
+  }
+
+  const priority = options?.priority ?? 'length';
+
+  if (priority === 'length') {
+    return Array.from({ length: options.length }, () => randElement(data));
+  }
+
+  const clonedData: T[] = JSON.parse(JSON.stringify(data));
+  const newArray: T[] = [];
+
+  while (clonedData.length && newArray.length !== options.length) {
+    const randomIndex = getRandomInRange({
+      min: 0,
+      max: clonedData.length - 1,
+    });
+    const item = clonedData[randomIndex];
+
+    newArray.push(item);
+    clonedData.splice(randomIndex, 1);
+  }
+
+  return newArray;
+}
+
+export const primitiveValueIsUnique: <T>(item: T, items: T[]) => boolean = (
```

Comment:
please remove to a new file `unique-validators.ts` and change the names to `dataUniqueValidator, ...`